lib/shared.bash

```diff
@@ -154,10 +154,17 @@ secret_exists() {
 secret_download() {
   local server="$1"
   local key="$2"
+
   if ! _secret=$(vault kv get -address="$server" -field=data -format=yaml "$key" | sed -r 's/: /=/; s/\"/\\"/g; s/\$/\\$/g; s/=(.*)$/=\"\1\"/g' ); then
-    echo "Failed to download secrets"
+    echo "Failed to download secrets: $_secret"
     exit 1
   fi
+
+  if [[ "${_secret:0:1}" == "{" ]]; then
+    # It's JSON, handle accordingly
+    _secret=$(vault kv get -address="$server" -field=data -format=json "$key" | sed -r 's/: /=/; s/\"/\\"/g;' )
+    _secret=$(jq -c 'walk(if type == "object" then with_entries(.key |= gsub("_"; "__") | gsub("\\."; "_")) else . end)' <<< "$_secret" | jq -r '[paths(scalars) as $p | {key: $p | join("_"), value: getpath($p)}] | .[] | "\(.key)=\(.value)"')
+  fi
```

Comment:
```suggestion
  if [[ "${_secret:0:1}" == "{" ]]; then
    # It's JSON, handle accordingly
    _secret=$(vault kv get -address="$server" -field=data -format=json "$key" | sed -r 's/: /=/; s/\"/\\"/g;' )
    _secret=$(jq -c 'walk(if type == "object" then with_entries(.key |= gsub("_"; "__") | gsub("\\."; "_")) else . end)' <<< "$_secret" | jq -r '[paths(scalars) as $p | {key: $p | join("_"), value: getpath($p)}] | .[] | "\(.key)=\(.value)"')
  fi
```

This was really hard for me to read and understand, maybe we should clean it up like:

```bash
if [[ "${_secret:0:1}" == "{" ]]; then
    # It's JSON, handle accordingly
    # Fetch the secret in JSON format
    _secret=$(vault kv get -address="$server" -field=data -format=json "$key")

    # Use jq to modify JSON keys and format into shell variables
    _secret=$(jq -c '
        walk(
            if type == "object" then 
                with_entries(
                    .key |= gsub("_"; "__") | gsub("\\."; "_")
                ) 
            else . end
        ) | 
        [paths(scalars) as $p | 
            {key: ($p | join("_")), value: getpath($p)}
        ] | 
        .[] | "\(.key)=\(.value)"' <<< "$_secret")
fi
```

Reply:
I had another thought, I wonder if we can inline this instead of making it two assignments to _secret:

```bash
    _secret=$(vault kv get -address="$server" -field=data -format=json "$key" | \
    jq -r '
        walk(
            if type == "object" then 
                with_entries(
                    .key |= gsub("_"; "__") | gsub("\\."; "_")
                ) 
            else . end
        ) | 
        [paths(scalars) as $p | 
            {key: ($p | join("_")), value: getpath($p)}
        ] | 
        .[] | "\(.key)=\(.value)"
    ')
    ```